subsys/dfu/dfu_multi_image/src/dfu_multi_image.c

```diff
@@ -176,6 +255,179 @@ static int process_current_item(const uint8_t *chunk, size_t chunk_size)
 	return chunk_size;
 }
 
+#ifdef CONFIG_DFU_MULTI_IMAGE_SAVE_PROGRESS
+
+static int module_settings_set(const char *key, size_t len_rd,
+			       settings_read_cb read_cb, void *cb_arg)
+{
+	ssize_t len;
+	int rc;
+
+	if (ctx.buffer != NULL && ctx.cur_image_no == IMAGE_NO_FIXED_HEADER
+	    && strcmp(key, CBOR_HEADER_SETTING_NAME) == 0) {
+		if (len_rd > (ctx.buffer_size - FIXED_HEADER_SIZE)) {
+			LOG_ERR("CBOR header in settings is too large");
+			return -ENOMEM;
+		}
+
+		/* Restore the cbor header from the NVS settings if available */
+		len = read_cb(cb_arg, ctx.buffer + FIXED_HEADER_SIZE,
+			      ctx.buffer_size - FIXED_HEADER_SIZE);
+		if (len < 0) {
+			LOG_ERR("Can't read cbor header from storage");
+			return len;
+		}
+
+		ctx.cur_item_size = FIXED_HEADER_SIZE + len;
+		ctx.cur_image_no = IMAGE_NO_CBOR_HEADER;
+
+		rc = parse_cbor_header();
+
+		if (rc < 0) {
+			LOG_ERR("Failed to parse cbor header loaded from storage: %d", rc);
+			return rc;
+		}
+
+		ctx.cur_offset = ctx.cur_item_size;
+
+		select_next_image();
+	}
+
+	if (ctx.buffer != NULL && strcmp(key, IMAGE_FINISHED_SETTING_NAME) == 0) {
+		/* Restore the cbor header from the NVS settings if available */
+		uint8_t finished_image_no = 0;
+
+		if (len_rd != sizeof(finished_image_no)) {
+			/* Should never get here */
+			return -EINVAL;
+		}
+
+		len = read_cb(cb_arg, &finished_image_no, sizeof(finished_image_no));
+
+		if (len < 0) {
+			LOG_ERR("Can't read finished image number from storage");
+			return len;
+		}
+
+		if (ctx.max_loaded_finished_image_no < finished_image_no) {
```

Comment:
This will not be true only if `settings_load` is called multiple times but in such a case it would probably be more efficient to return before `read_cb` for the second load?

Reply:
Ok, removed this